lib/shared/open_ai/api_client.dart

```diff
@@ -0,0 +1,369 @@
+                                  },
+                                  'duration': {
+                                    'type': 'integer',
+                                    'description': 'Duration in minutes'
+                                  }
+                                },
+                                'required': [
+                                  'stepNumber',
+                                  'instruction',
+                                  'duration'
+                                ]
+                              }
+                            }
+                          },
+                          'required': [
+                            'name',
+                            'description',
+                            'ingredients',
+                            'steps'
+                          ]
+                        }
+                      },
+                      'required': ['dayNumber', 'recipe']
+                    }
+                  }
+                },
+                'required': ['entries']
+              }
+            },
+            'required': ['mealPlan']
+          },
+        ),
+      ],
+      functionCall: {'name': 'generate_meal_plan'},
+      maxTokens: 3000,
+    );
+
+    try {
+      final response = await _makeApiCall(requestBody, user.apiKey!);
+      if (response != null) {
+        final existingIngredients = await isar.ingredients.where().findAll();
+
+        final functionCall = response['choices'][0]['message']['function_call'];
+        final arguments = jsonDecode(functionCall['arguments']);
+
+        MealPlan mealPlan = await isar.writeTxn(() async {
+          final mealPlan = MealPlan()
+            ..startDate = DateTime.now()
+            ..endDate = DateTime.now().add(Duration(days: 5));
+          isar.mealPlans.put(mealPlan);
+
+          for (var entryData in arguments['mealPlan']['entries']) {
+            final mealPlanEntry = MealPlanEntry()
+              ..day = DateTime.now()
+                  .add(Duration(days: entryData['dayNumber'] - 1));
+
+            final recipeData = entryData['recipe'];
+            final recipe = Recipe()
+              ..title = recipeData['name']
+              ..description = recipeData['description']
+              ..cookingDuration = recipeData['steps'].fold<int>(
+                  0, (int sum, dynamic step) => sum + step['duration'] as int);
+
+            await isar.mealPlanEntrys.put(mealPlanEntry);
+            await isar.recipes.put(recipe);
+            mealPlanEntry.recipe.add(recipe);
+            await mealPlanEntry.recipe.save();
+            mealPlanEntry.mealPlan.add(mealPlan);
+            await mealPlanEntry.mealPlan.save();
+
+            for (var ingredientData in recipeData['ingredients']) {
+              final ingredient = await _findOrCreateIngredient(
+                  ingredientData, existingIngredients, isar);
+
+              final recipeIngredient = RecipeIngredient()
```

Comment:
Vielleicht kann man hier die einzelne Schleifen in eigene Methoden tun um das lesbarerer zu machen. Ist schon sehr viel hier